File: tests/runners/csharp/tests/Test_10_pack_embed.cs
// Test Case 10: Pack Embed
// Tests @pack annotation on embed types for serializing fields to a single string

using System;

class Program
{
    static int passed = 0;
    static int failed = 0;

    static void Assert(bool condition, string message)
    {
        if (!condition)
        {
            Console.WriteLine($"    FAILED: {message}");
            failed++;
        }
    }

    static void TestPositionPack()
    {
        Console.WriteLine("  Testing Position pack/unpack (sep: ;)...");

        var pos = new test.pack_embed.Position { x = 100.5f, y = 200.3f };
        var packed = pos.Pack();
        Assert(packed == "100.5;200.3", $"Position pack: expected '100.5;200.3' got '{packed}'");

        var unpacked = test.pack_embed.Position.Unpack(packed);
        Assert(Math.Abs(unpacked.x - 100.5f) < 0.01f, "Position unpack x");
        Assert(Math.Abs(unpacked.y - 200.3f) < 0.01f, "Position unpack y");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestPosition3DPack()
    {
        Console.WriteLine("  Testing Position3D pack/unpack (sep: ;)...");

        var pos = new test.pack_embed.Position3D { x = 10f, y = 20f, z = 30f };
        var packed = pos.Pack();
        Assert(packed == "10;20;30", $"Position3D pack: expected '10;20;30' got '{packed}'");

        var unpacked = test.pack_embed.Position3D.Unpack(packed);
        Assert(Math.Abs(unpacked.x - 10f) < 0.01f, "Position3D unpack x");
        Assert(Math.Abs(unpacked.y - 20f) < 0.01f, "Position3D unpack y");
        Assert(Math.Abs(unpacked.z - 30f) < 0.01f, "Position3D unpack z");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestColorPack()
    {
        Console.WriteLine("  Testing Color pack/unpack (sep: ,)...");

        var c = new test.pack_embed.Color { r = 255, g = 128, b = 64 };
        var packed = c.Pack();
        Assert(packed == "255,128,64", $"Color pack: expected '255,128,64' got '{packed}'");

        var unpacked = test.pack_embed.Color.Unpack(packed);
        Assert(unpacked.r == 255, "Color unpack r");
        Assert(unpacked.g == 128, "Color unpack g");
        Assert(unpacked.b == 64, "Color unpack b");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestColorAlphaPack()
    {
        Console.WriteLine("  Testing ColorAlpha pack/unpack (sep: |)...");

        var ca = new test.pack_embed.ColorAlpha { r = 255, g = 255, b = 255, a = 128 };
        var packed = ca.Pack();
        Assert(packed == "255|255|255|128", $"ColorAlpha pack: expected '255|255|255|128' got '{packed}'");

        var unpacked = test.pack_embed.ColorAlpha.Unpack(packed);
        Assert(unpacked.r == 255, "ColorAlpha unpack r");
        Assert(unpacked.g == 255, "ColorAlpha unpack g");
        Assert(unpacked.b == 255, "ColorAlpha unpack b");
        Assert(unpacked.a == 128, "ColorAlpha unpack a");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestSizePack()
    {
        Console.WriteLine("  Testing Size pack/unpack (sep: ;)...");

        var s = new test.pack_embed.Size { width = 800, height = 600 };
        var packed = s.Pack();
        Assert(packed == "800;600", $"Size pack: expected '800;600' got '{packed}'");

        var unpacked = test.pack_embed.Size.Unpack(packed);
        Assert(unpacked.width == 800, "Size unpack width");
        Assert(unpacked.height == 600, "Size unpack height");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestRangePack()
    {
        Console.WriteLine("  Testing Range pack/unpack (sep: ~)...");

        var r = new test.pack_embed.Range { min = -100, max = 100 };
        var packed = r.Pack();
        Assert(packed == "-100~100", $"Range pack: expected '-100~100' got '{packed}'");

        var unpacked = test.pack_embed.Range.Unpack(packed);
        Assert(unpacked.min == -100, "Range unpack min");
        Assert(unpacked.max == 100, "Range unpack max");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestTryUnpack()
    {
        Console.WriteLine("  Testing TryUnpack failure cases...");

        bool ok = test.pack_embed.Position.TryUnpack("invalid", out var result);
        Assert(!ok, "TryUnpack should fail on invalid input");

        ok = test.pack_embed.Position.TryUnpack("1.0;2.0", out result);
        Assert(ok, "TryUnpack should succeed on valid input");
        Assert(Math.Abs(result.x - 1.0f) < 0.01f, "TryUnpack x");
        Assert(Math.Abs(result.y - 2.0f) < 0.01f, "TryUnpack y");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestStatsNoPack()
    {
        Console.WriteLine("  Testing Stats (no @pack) works as normal embed...");

        var stats = new test.pack_embed.Stats { hp = 100, mp = 50, attack = 25, defense = 10 };
        Assert(stats.hp == 100, "Stats hp");
        Assert(stats.mp == 50, "Stats mp");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void Main(string[] args)
    {
        Console.WriteLine("=== Test Case 10: Pack Embed ===");

        TestPositionPack();
        TestPosition3DPack();
        TestColorPack();
        TestColorAlphaPack();
        TestSizePack();
        TestRangePack();
        TestTryUnpack();
        TestStatsNoPack();

        Console.WriteLine($"\nResults: {passed} passed, {failed} failed");
        if (failed > 0)
        {
            Environment.Exit(1);
        }
        Console.WriteLine("=== All tests passed! ===");
    }
}

File: tests/runners/rust/tests/test_10_pack_embed.rs
// Test Case 10: Pack Embed
// Tests @pack annotation on embed types for serializing fields to a single string

use polygen_test::schema::test_pack_embed::*;

fn main() {
    println!("=== Test Case 10: Pack Embed ===");

    test_position_pack();
    test_position3d_pack();
    test_color_pack();
    test_color_alpha_pack();
    test_size_pack();
    test_range_pack();
    test_unpack_error();
    test_stats_no_pack();

    println!("=== All tests passed! ===");
}

fn test_position_pack() {
    println!("  Testing Position pack/unpack (sep: ;)...");

    let pos = Position { x: 100.5, y: 200.3 };
    let packed = pos.pack();
    assert_eq!(packed, "100.5;200.3");

    let unpacked = Position::unpack(&packed).unwrap();
    assert!((unpacked.x - 100.5).abs() < 0.01);
    assert!((unpacked.y - 200.3).abs() < 0.01);

    println!("    PASS");
}

fn test_position3d_pack() {
    println!("  Testing Position3D pack/unpack (sep: ;)...");

    let pos = Position3D { x: 10.0, y: 20.0, z: 30.0 };
    let packed = pos.pack();
    assert_eq!(packed, "10;20;30");

    let unpacked = Position3D::unpack(&packed).unwrap();
    assert!((unpacked.x - 10.0).abs() < 0.01);
    assert!((unpacked.y - 20.0).abs() < 0.01);
    assert!((unpacked.z - 30.0).abs() < 0.01);

    println!("    PASS");
}

fn test_color_pack() {
    println!("  Testing Color pack/unpack (sep: ,)...");

    let c = Color { r: 255, g: 128, b: 64 };
    let packed = c.pack();
    assert_eq!(packed, "255,128,64");

    let unpacked = Color::unpack(&packed).unwrap();
    assert_eq!(unpacked.r, 255);
    assert_eq!(unpacked.g, 128);
    assert_eq!(unpacked.b, 64);

    println!("    PASS");
}

fn test_color_alpha_pack() {
    println!("  Testing ColorAlpha pack/unpack (sep: |)...");

    let ca = ColorAlpha { r: 255, g: 255, b: 255, a: 128 };
    let packed = ca.pack();
    assert_eq!(packed, "255|255|255|128");

    let unpacked = ColorAlpha::unpack(&packed).unwrap();
    assert_eq!(unpacked.r, 255);
    assert_eq!(unpacked.g, 255);
    assert_eq!(unpacked.b, 255);
    assert_eq!(unpacked.a, 128);

    println!("    PASS");
}

fn test_size_pack() {
    println!("  Testing Size pack/unpack (sep: ;)...");

    let s = Size { width: 800, height: 600 };
    let packed = s.pack();
    assert_eq!(packed, "800;600");

    let unpacked = Size::unpack(&packed).unwrap();
    assert_eq!(unpacked.width, 800);
    assert_eq!(unpacked.height, 600);

    println!("    PASS");
}

fn test_range_pack() {
    println!("  Testing Range pack/unpack (sep: ~)...");

    let r = Range { min: -100, max: 100 };
    let packed = r.pack();
    assert_eq!(packed, "-100~100");

    let unpacked = Range::unpack(&packed).unwrap();
    assert_eq!(unpacked.min, -100);
    assert_eq!(unpacked.max, 100);

    println!("    PASS");
}

fn test_unpack_error() {
    println!("  Testing unpack error cases...");

    let result = Position::unpack("invalid");
    assert!(result.is_err());

    let result = Position::unpack("1;2;3");
    assert!(result.is_err());

    println!("    PASS");
}

fn test_stats_no_pack() {
    println!("  Testing Stats (no @pack) works as normal embed...");

    let stats = Stats {
        hp: 100,
        mp: 50,
        attack: 25,
        defense: 10,
    };

    assert_eq!(stats.hp, 100);
    assert_eq!(stats.mp, 50);

    println!("    PASS");
}
